pkg/reconciler/apis/apibinding/apibinding_reconcile.go

```diff
@@ -421,6 +422,7 @@ func crdFromAPIResourceSchema(schema *apisv1alpha1.APIResourceSchema) (*apiexten
 				annotationBoundCRDKey:      "",
 				annotationSchemaClusterKey: schema.ClusterName,
 				annotationSchemaNameKey:    schema.Name,
+				annotationAPIIdentityKey:   hash,
```

Comment:
this will mean two schemas with different identities (can happen with multiple APIExports) will fight.

Reply:
Thanks, will remove

Reply:
I knew I wanted to come back to this! In https://github.com/kcp-dev/kubernetes/pull/57/files that we just merged, it uses this annotation to alter the resource prefix of the CRs for the CRD in etcd.

I'm guessing you're thinking of a scenario like this?

- APIExport E1 exports APIResourceSchemas S1, S2
- APIExport E2 exports S1, S2' (a variant of S2)
- User creates APIBinding B1 in ws W1 pointing to E1
    - This creates CRD C1 in system:bound-crds with name=S1.metadata.UID, identityAnnotation=E1.identityHash
    - This creates CRD C2 in system:bound-crds with name=S2.metadata.UID, identityAnnotation=E1.identityHash
- User creates APIBinding B2 in ws W2 pointing to E2
    - Because C1 already exists (for S1), we don't try to create a new CRD (even though E2 has a different identityHash)
    - This creates CRD C3 in system:bound-crds with name=S2'.metadata.UID, identityAnnotation=E2.identityHash

What this tells me is that bound CRDs can't be associated back to identity hashes.

So the question becomes: how do we specify a schema's identity for storage in etcd, for instances of said schema? It's more like each APIResourceSchema needs a separate associated APIIdentity object (?) to represent compatibility and to be able to distinguish when there are breaking changes. For example, S2 and S2' could both share an APIIdentity if their actual schemas are compatible (e.g. S2' adds a new optional field). That isn't tied to the APIExport per se.

Thoughts?

Reply:
Chatted on Slack a few days ago. When getting a CRD for a specific known identity, we'll add the annotation on the fly (not stored in etcd). And the identity is via the APIExport, not a separate resource.